### ISSUE TITLE
Repeating EliminateEssentialBC

### ISSUE THREAD

**Issue by Watchmojo** [mfem/mfem#36]
Hi,

Sorry for so many questions:) Please take your time to answer my questions. 

For my problem, I need to solve Ax=b repeatedly. Fortunately, my A does not change, so A can be assembled only once. However, vector b needs to be updated repeatedly. 

My question is how to do "a->EliminateEssentialBC(ess_bdr, x, *b)" repeatedly without destroying ParBilinearForm  a? Because constructing ParBilinearForm in every loop is only a waste of time.

The same problem for GetDerivate function: why calculating dshape and inv_jac for every call? why not save these matrix for future use?

Thank you very much for your patience. 
Frank

**Comment by tzanio** [comment]
Hi Frank,

If you have zero boundary conditions, than you can reuse the eliminated matrix. In general however, you are correct that in v3.0 you will need to repeatedly assemble the matrix.

We have added a new functionality in the next release, v3.1, to address this issue. I suggest that you use rebuilding as a temporary solution, and then switch to the new interface when the release comes out in the next couple of weeks.

Hope this helps,
Tzanio

**Comment by Watchmojo** [comment]
Hi, Tzanio,

Okay. I will follow your suggestion. 

I also noticed somethings that can be improved/added in MFEM. for example:

1. The ability to restart from the previous calculation. To do this, gridfunction should be able to write and read into a binary format file. 
2. Tecplot format can be supported in datacollection class. 
3. BLAS can be used to further optimize vector class. 

 When I am available, I will start working on these to contribute to MFEM.

Looking forward to MFEM-3.1 and thank you for your work.
Frank

**Comment by tzanio** [comment]
Hi Frank,

These all sound good, and will be nice to have in the library. Thank you for your interest in contributing. Let us know if we can help.

Tzanio

**Comment by martemyev** [comment]
Hi all,

I have a follow up comment/question on a similar problem that I've recently encountered myself.
@Watchmojo I believe that you can avoid reassembling matrices even if you have non-zero Dirichlet b.c., but you can't use EliminateEssentialBC for that, because it calls EliminateRowCol function which gets rid of both row and column corresponding to the boundary dof, and eliminating a column affects the RHS vector w.r.t. to the values of the b.c. This maintains the symmetry of the matrix, and allows the usage of the CG solver. However, if you eliminate only rows of the matrices, that doesn't affect RHS, and you can eliminate b.c. only once for a given matrix, while solving SLAEs with non-symmetric matrix for different RHS without reassembling and re-eliminating b.c. from the matrix.
@tzanio If I am right about it, is it possible to compute an approximate of the inverse matrix before the solution of SLAEs with multiple RHS to speed it up? I'm thinking about a preconditioner computed in advance, that can be used later in the iterative solvers. Would it work? Can it be done with the current MFEM code?

Thanks,
Mikhail

**Comment by martemyev** [comment]
Hi all,

Here, I post some working code to prove my previous point for a problem, when a 3D linear elasticity problem with zero RHS is solved and the Dirichlet b.c. are all zero except one which is equal to one at each boundary dof for each separate solution.

`#include "mfem.hpp"
#include <cmath>
#include <vector>

void get_bdr_vdofs(const mfem::FiniteElementSpace& fespace,
                   const mfem::Array<int>& ess_bdr,
                   std::vector<int>& bdr_vdofs)
{
  mfem::Array<int> ess_vdofs;
  fespace.GetEssentialVDofs(ess_bdr, ess_vdofs); // bdr dofs are marked by -1
  bdr_vdofs.clear();
  bdr_vdofs.reserve(ess_vdofs.Size());
  for (int i = 0; i < ess_vdofs.Size(); ++i) {
    if (ess_vdofs[i] < 0)
      bdr_vdofs.push_back(i);
  }
}

void test_snapshots()
{
  const int ncells = 6;
  const int gen_edges = 1;
  const double size = 1000.;
  mfem::Mesh mesh(ncells, ncells, ncells, mfem::Element::HEXAHEDRON, gen_edges,
                  size);
  const int dim = mesh.Dimension();

  const int order = 1;
  mfem::H1_FECollection fec(order, dim);
  mfem::FiniteElementSpace fespace(&mesh, &fec, dim);

  mfem::Array<int> ess_bdr(mesh.bdr_attributes.Max());
  ess_bdr = 1;
  std::vector<int> bdr_vdofs;
  get_bdr_vdofs(fespace, ess_bdr, bdr_vdofs);

  std::cout << "n bdr vdofs " << bdr_vdofs.size() << std::endl;

  std::vector<mfem::Vector> solutions1(bdr_vdofs.size());
  std::vector<mfem::Vector> solutions2(bdr_vdofs.size());

  const double rho = 2500.;
  const double vp  = 3500.;
  const double vs  = 2000.;
  mfem::ConstantCoefficient lambda(rho*(vp*vp-2.*vs*vs));
  mfem::ConstantCoefficient mu(rho*vs*vs);

  mfem::Vector F(fespace.GetVSize());

  int maxiter = 1000;
  double rtol = 1e-24;
  double atol = 1e-48;
  bool use_default_solver_options = true;

  // approach 1
  mfem::StopWatch timer;
  timer.Start();
  mfem::Vector bdr_values(fespace.GetVSize());
  if (use_default_solver_options) {
    maxiter = 1000;
    rtol    = 1e-12;
    atol    = 1e-24;
  }
  for (size_t bd = 0; bd < bdr_vdofs.size(); ++bd) {
    mfem::BilinearForm stif(&fespace);
    stif.AddDomainIntegrator(new mfem::ElasticityIntegrator(lambda, mu));
    stif.Assemble();

    bdr_values = 0.;
    bdr_values(bdr_vdofs[bd]) = 1.;

    F = 0.;
    stif.EliminateEssentialBC(ess_bdr, bdr_values, F);

    stif.Finalize();
    const mfem::SparseMatrix& S = stif.SpMat();

    mfem::GSSmoother precond(S);
    solutions1[bd].SetSize(fespace.GetVSize());
    solutions1[bd] = 0.;
    mfem::PCG(S, precond, F, solutions1[bd], 0, maxiter, rtol, atol);
  }
  timer.Stop();
  std::cout << "approach 1 took " << timer.UserTime() << " sec" << std::endl;

  // approach 2
  timer.Clear();
  timer.Start();
  mfem::BilinearForm stif(&fespace);
  stif.AddDomainIntegrator(new mfem::ElasticityIntegrator(lambda, mu));
  stif.Assemble();
  {
    mfem::SparseMatrix& S = stif.SpMat();
    for (size_t bd = 0; bd < bdr_vdofs.size(); ++bd)
      S.EliminateRow(bdr_vdofs[bd], 1);
  }
  stif.Finalize();
  const mfem::SparseMatrix& S = stif.SpMat();
  if (use_default_solver_options) {
    maxiter = 10;
    rtol    = 0.;
    atol    = 0.;
  }
  mfem::GSSmoother precond(S);
  mfem::GMRESSolver solver;
  solver.SetPreconditioner(precond);
  solver.SetOperator(S);
  solver.SetMaxIter(maxiter);
  solver.SetRelTol(rtol);
  solver.SetAbsTol(atol);
  for (size_t bd = 0; bd < bdr_vdofs.size(); ++bd) {
    F = 0.;
    F(bdr_vdofs[bd]) = 1.;

    solutions2[bd].SetSize(fespace.GetVSize());
    solutions2[bd] = 0.;
    solver.Mult(F, solutions2[bd]);
  }
  timer.Stop();
  std::cout << "approach 2 took " << timer.UserTime() << " sec" << std::endl;

  // comparison
  double max_abs_diff = 0.;
  double max_rel_diff = 0.;
  int max_abs_pos[] = { -1, -1 };
  int max_rel_pos[] = { -1, -1 };
  for (size_t bd = 0; bd < bdr_vdofs.size(); ++bd) {
    for (int i = 0; i < fespace.GetVSize(); ++i) {
      double diff = fabs(solutions1[bd](i) - solutions2[bd](i));
      if (diff > max_abs_diff) {
        max_abs_diff = diff;
        max_abs_pos[0] = bd;
        max_abs_pos[1] = i;
      }
      if (fabs(solutions1[bd](i)) > 1e-8)
        diff /= fabs(solutions1[bd](i));
      if (diff > max_rel_diff) {
        max_rel_diff = diff;
        max_rel_pos[0] = bd;
        max_rel_pos[1] = i;
      }
    }
  }
  std::cout << "max abs diff " << max_abs_diff
            << " bd " << max_abs_pos[0]
            << " el " << max_abs_pos[1]
            << " sol1 " << solutions1[max_abs_pos[0]](max_abs_pos[1])
            << " sol2 " << solutions2[max_abs_pos[0]](max_abs_pos[1])
            << std::endl;
  std::cout << "max rel diff " << max_rel_diff
            << " bd " << max_rel_pos[0]
            << " el " << max_rel_pos[1]
            << " sol1 " << solutions1[max_rel_pos[0]](max_rel_pos[1])
            << " sol2 " << solutions2[max_rel_pos[0]](max_rel_pos[1])
            << std::endl;
}
`

**Comment by tzanio** [comment]
Frank, to answer your original question:

The new functions [BilinearForm::FormLinearSystem](http://mfem.github.io/doxygen/html/classmfem_1_1BilinearForm.html#a476580737152dab7dad2aa3d82ac2248) and  [ParBilinearForm::FormLinearSystem](http://mfem.github.io/doxygen/html/classmfem_1_1ParBilinearForm.html#ab7b19b4295a780b5c7e23bd98ec997d9) (in mfem-3.1) can be called repeatedly with different righ-hand sides. We don't have such an example yet, but you can see how to use `FormLinearSystem` in Examples 1-4.

**Comment by tzanio** [comment]
Hi Mikhail,

Can you try re-running your tests with the `FormLinearSystem` method above?

This method saves the eliminated parts of the matrix, so modifying the right-hand does not involve re-assembly but a matvec with `Aib`, the `interior x boundary` sub-matrix. I suspect that this will be faster than eliminating only the rows, since `FormLinearSystem` will multiply with `Aib`, while the second approach will multiply with this sub-matrix on each iteration. PCG also may take less iterations than GMRES (especially with a small restart parameter).

Thank you for the interesting suggestion. Please let us know what you find.
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,
Congrats on the new MFEM release!
I've tried the new approach on the previous test, and indeed, it's faster than the first two. Here is the relevant code:
```cxx
mfem::Array<int> ess_tdof_list; // this is analogous to bdr_vdofs
  if (mesh.bdr_attributes.Size()) {
    mfem::Array<int> ess_bdr(mesh.bdr_attributes.Max());
    ess_bdr = 1;
    fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
  }
  std::vector<mfem::Vector> solutions3(ess_tdof_list.Size());

  // approach 3
  timer.Clear();
  timer.Start();
  mfem::BilinearForm Stif(&fespace);
  Stif.AddDomainIntegrator(new mfem::ElasticityIntegrator(lambda, mu));
  Stif.Assemble();
  mfem::Vector x(fespace.GetVSize()); // boundary values
  mfem::Vector b(fespace.GetVSize()); // RHS (it's always 0 in the loop)
  if (use_default_solver_options) {
    maxiter = 1000;
    rtol    = 1e-12;
    atol    = 1e-24;
  }
  for (int bd = 0; bd < ess_tdof_list.Size(); ++bd) {
    x = 0.;
    x(ess_tdof_list[bd]) = 1.;
    b = 0.;

    mfem::SparseMatrix A; // mat after eliminating b.c.
    mfem::Vector X, B;
    Stif.FormLinearSystem(ess_tdof_list, x, b, A, X, B);

    mfem::GSSmoother precond(A);
    mfem::PCG(A, precond, B, X, 0, maxiter, rtol, atol);
    solutions3[bd] = X;
  }
  timer.Stop();
  std::cout << "approach 3 took " << timer.UserTime() << " sec" << std::endl;
```
The results obtained in comparison to the first two approaches with `use_default_solver_options = false` are the following (note that I ran the new tests on a different machine, so the time is different for the previous approaches, but what is relevant here is the comparison between times)
```
n bdr vdofs 654
approach 1 took 10.7587 sec
approach 2 took 5.5216 sec
approach 3 took 1.04756 sec

comparison of approaches 1 and 2
max abs diff 5.06755e-11 bd 216 el 341 sol1 1 sol2 1
max rel diff 1.89268e-05 bd 319 el 807 sol1 -1.01835e-08 sol2 -1.01837e-08

comparison of approaches 1 and 3
max abs diff 5.06755e-11 bd 216 el 341 sol1 1 sol2 1
max rel diff 1.01751e-09 bd 337 el 823 sol1 -1.0799e-08 sol2 -1.0799e-08

comparison of approaches 2 and 3
max abs diff 2.36375e-13 bd 334 el 901 sol1 6.37669e-09 sol2 6.37693e-09
max rel diff 1.89267e-05 bd 319 el 807 sol1 -1.01837e-08 sol2 -1.01835e-08
```
Also, with `use_default_solver_options = true`, I compared the approaches 1 and 3, because they both use `mfem::PCG` as a solver, and here is the result:
```
approach 1 took 9.64295 sec
approach 3 took 0.562211 sec

comparison of approaches 1 and 3
max abs diff 7.57261e-05 bd 202 el 327 sol1 0.999924 sol2 1
max rel diff 9.83025e-05 bd 579 el 939 sol1 3.09822e-13 sol2 3.09853e-13
```
So, the conclusion is obvious - approach3 is great.

The only thing that I didn't comprehend is when I use the following code for the approach3:
```
for (int bd = 0; bd < ess_tdof_list.Size(); ++bd) {
    x = 0.;
    x(ess_tdof_list[bd]) = 1.;
    b = 0.;

    mfem::SparseMatrix A; // mat after eliminating b.c.
    mfem::Vector B;
    Stif.FormLinearSystem(ess_tdof_list, x, b, A, solutions3[bd], B);

    mfem::GSSmoother precond(A);
    mfem::PCG(A, precond, B, solutions3[bd], 0, maxiter, rtol, atol);
  }
```
i.e. when I substitute the usage of the vector `X` in the loop by the `solutions3[bd]` directly to avoid extra copying `solutions3[bd] = X;`, the code doesn't work.
Here is a simple test for a mesh consisting of one element, i.e. with all the solutions governed by the boundary conditions:
```cxx
#include "mfem.hpp"
#include <cmath>
#include <vector>

void print(const std::vector<mfem::Vector> &solutions)
{
  for (size_t i = 0; i < solutions.size(); ++i) {
    for (int j = 0; j < solutions[i].Size(); ++j) {
      std::cout << solutions[i](j) << " ";
    }
    std::cout << std::endl;
  }
}

void test_snapshots()
{
  const int ncells = 1;
  const int gen_edges = 1;
  const double size = 1000.;
  mfem::Mesh mesh(ncells, ncells, ncells, mfem::Element::HEXAHEDRON, gen_edges, size);
  const int dim = mesh.Dimension();

  const int order = 1;
  mfem::H1_FECollection fec(order, dim);
  mfem::FiniteElementSpace fespace(&mesh, &fec, dim);

  const double rho = 2500.;
  const double vp  = 3500.;
  const double vs  = 2000.;
  mfem::ConstantCoefficient lambda(rho*(vp*vp-2.*vs*vs));
  mfem::ConstantCoefficient mu(rho*vs*vs);

  int maxiter = 1000;
  double rtol = 1e-12;
  double atol = 1e-24;

  mfem::Array<int> ess_tdof_list;
  if (mesh.bdr_attributes.Size()) {
    mfem::Array<int> ess_bdr(mesh.bdr_attributes.Max());
    ess_bdr = 1;
    fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
  }
  std::vector<mfem::Vector> solutions3_a(ess_tdof_list.Size());
  std::vector<mfem::Vector> solutions3_b(ess_tdof_list.Size());

  mfem::BilinearForm Stif(&fespace);
  Stif.AddDomainIntegrator(new mfem::ElasticityIntegrator(lambda, mu));
  Stif.Assemble();
  mfem::Vector x(fespace.GetVSize()); // boundary values
  mfem::Vector b(fespace.GetVSize()); // RHS (it's always 0 in the loop)

  for (int bd = 0; bd < ess_tdof_list.Size(); ++bd) {
    x = 0.;
    x(ess_tdof_list[bd]) = 1.;
    b = 0.;

    mfem::SparseMatrix A;
    mfem::Vector X, B;
    Stif.FormLinearSystem(ess_tdof_list, x, b, A, X, B);

    mfem::GSSmoother precond(A);
    mfem::PCG(A, precond, B, X, 0, maxiter, rtol, atol);
    solutions3_a[bd] = X;
  }

  for (int bd = 0; bd < ess_tdof_list.Size(); ++bd) {
    x = 0.;
    x(ess_tdof_list[bd]) = 1.;
    b = 0.;

    mfem::SparseMatrix A;
    mfem::Vector B;
    Stif.FormLinearSystem(ess_tdof_list, x, b, A, solutions3_b[bd], B);

    mfem::GSSmoother precond(A);
    mfem::PCG(A, precond, B, solutions3_b[bd], 0, maxiter, rtol, atol);
  }

  std::cout << "Using X\n";
  print(solutions3_a);
  std::cout << "Using solutions3 directly\n";
  print(solutions3_b);
}
```
The output is
```
Using X
1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
Using solutions3 directly
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 
```
So, I'm really not sure what causes this behavior.

Thanks,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

The correct code for approach 3 will be something like this:

```c++
  for (int bd = 0; bd < ess_tdof_list.Size(); ++bd) {
    mfem::Vector &x = solutions3_b[bd];
    x.SetSize(fespace.GetVSize());
    x = 0.;
    x(ess_tdof_list[bd]) = 1.;
    b = 0.;

    mfem::SparseMatrix A;
    mfem::Vector B, X;
    Stif.FormLinearSystem(ess_tdof_list, x, b, A, X, B);

    mfem::GSSmoother precond(A);
    mfem::PCG(A, precond, B, X, 0, maxiter, rtol, atol);

    Stif.RecoverFEMSolution(X, b, x);
 }
```

If `X` and `x` are of the same size, there is no copying (the data is reused), which was the problem with the original code. Sorry if this was not clear, we will clarify it in the comments.

The above is also more general, as it will work also when `X` and `x` have different sizes, e.g. if you are using AMR or static condensation.

Hope this helps,
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,

Now I see how to use it. Thanks a lot!

Cheers,
Mikhail
